### PR TITLE
lint: enable staticcheck

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

Follow-up to: https://github.com/osmosis-labs/osmosis/issues/1897

Please note that this is a `staticcheck` set of rules (a subset of rules from `staticcheck` binary

## Brief Changelog

- Removed variables that were not being reused after assignment
- Handled errors from `defer Close()`
   * More context: https://www.joeshaw.org/dont-defer-close-on-writable-files/
- Swapped deprecated `GenerateSaveCoinKey`
- Removed redundant check against 0 for `uint`s

## Testing and Verifying

This change is a trivial rework / code cleanup without any test coverage.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/lockup/types/genesis.go [hunk 1]
```diff
@@ -13,8 +11,5 @@ func DefaultGenesis() *GenesisState {
 // Validate performs basic genesis state validation returning an error upon any
 // failure.
 func (gs GenesisState) Validate() error {
-	if gs.LastLockId < 0 {
```
Comment (alexanderbez) [inline x/lockup/types/genesis.go:16]: Why was this removed?

File: x/incentives/types/genesis.go [hunk 2]
```diff
@@ -45,8 +45,5 @@ func (gs GenesisState) Validate() error {
 	if gs.Params.DistrEpochIdentifier == "" {
 		return errors.New("epoch identifier should NOT be empty")
 	}
-	if gs.LastGaugeId < 0 {
```
Comment (alexanderbez) [inline x/incentives/types/genesis.go:48]: ditto

File: cmd/osmosisd/cmd/balances_from_state_export.go [hunk 3]
```diff
@@ -179,11 +179,6 @@ Example:
 
 			authGenesis := authtypes.GenesisState{}
 			clientCtx.Codec.MustUnmarshalJSON(genState["auth"], &authGenesis)
-			accounts, err := authtypes.UnpackAccounts(authGenesis.Accounts)
```
Comment (alexanderbez) [inline cmd/osmosisd/cmd/balances_from_state_export.go:182]: Was this removed on purpose?

File: x/lockup/types/genesis.go (hunk 1 above)
Comment (p0mvn) [reply x/lockup/types/genesis.go:16]: This is a `uint` so it cannot be less than 0

File: x/incentives/types/genesis.go (hunk 2 above)
Comment (p0mvn) [reply x/incentives/types/genesis.go:48]: Same reason - `uint`

File: cmd/osmosisd/cmd/balances_from_state_export.go (hunk 3 above)
Comment (p0mvn) [reply cmd/osmosisd/cmd/balances_from_state_export.go:182]: Yes, this was unused other than unpacking

Comment (p0mvn) [reply cmd/osmosisd/cmd/balances_from_state_export.go:182]: I just double-checked, and I think the two lines above can also be removed:
```
authGenesis := authtypes.GenesisState{}
clientCtx.Codec.MustUnmarshalJSON(genState["auth"], &authGenesis)
```

Generally, I'm having difficulty understanding what's happening in that command. We might want to consider having an issue to determine what is the need for it and potentially refactor

File: x/epochs/handler.go [hunk 4]
```diff
@@ -13,8 +13,6 @@ import (
 // NewHandler returns a handler for epochs module messages.
 func NewHandler(k keeper.Keeper) sdk.Handler {
 	return func(ctx sdk.Context, msg sdk.Msg) (*sdk.Result, error) {
-		ctx = ctx.WithEventManager(sdk.NewEventManager())
```
Comment (alexanderbez) [inline x/epochs/handler.go:16]: IMO we shouldn't even have these `handler.go` files -- it's all legacy code.

File: x/incentives/types/genesis.go (hunk 2 above)
Comment (alexanderbez) [reply x/incentives/types/genesis.go:48]: ACK

**Review by alexanderbez — APPROVED** [review]

File: x/epochs/handler.go (hunk 4 above)
Comment (p0mvn) [reply x/epochs/handler.go:16]: I'm guessing this: https://github.com/cosmos/cosmos-sdk/pull/9650 would be the right way to do it?

File: cmd/osmosisd/cmd/forceprune.go [hunk 5]
```diff
@@ -128,13 +139,21 @@ func compactBlockStore(dbPath string) error {
...
 	if err = db.CompactRange(*util.BytesPrefix([]byte{})); err != nil {
 		return err
 	}
+	// N.B: We duplicate the call to db.Close() on top of
+	// the call in defer statement above to make sure that the resources
+	// are properly released and any potential error from Close()
+	// is handled. Close() should be idempotent so this is acceptable.
+	if err := db.Close(); err != nil {
+		return err
+	}
```
Comment (ValarDragon) [inline cmd/osmosisd/cmd/forceprune.go:156]: I don't get why are we doing this? The defer handles this more cleanly?

Comment (p0mvn) [reply cmd/osmosisd/cmd/forceprune.go:156]: If we do `defer db.Close()`, it does not do error handling.

The new change validates the error while keeping `defer db.Close()` as well. The code in `defer` runs even if the code panics, making sure that we always close the resources. Calling this twice should not be a problem because `Close()` is idempotent

Comment (ValarDragon) [reply cmd/osmosisd/cmd/forceprune.go:156]: Can we instead do a named return error, (e.g. `func compactBlockStore(dbPath string) (rerr error)`, and then do `defer rerr = db.Close()`. This is the general pattern for handling errors in a panic.

Encouraging db closing being isolated scope / not handled multiple times is ideal imo.

**Review by ValarDragon — APPROVED** [review]
LGTM after we resolve the db.Close

**Comment by p0mvn** [comment]
Merging since db.Close addressed + 2 approvals
